### ISSUE TITLE
Support C-style enums

### ISSUE THREAD

**Issue by jgalenson** [dtolnay/cxx#132]
It would be nice to support C-style enums both from Rust to C++ and C++ to Rust.

The former is straightforward, but for the latter we must handle C++ code that passes values outside the expected range (which is illegal in Rust).  We can do this by auto-generating a conversion function that checks whether the value from C++ is one of the expected values and aborting otherwise.

The #[non_exhaustive] attribute is a possibility instead of a conversion function, but I believe it would not help at the moment (as both the definition and uses are in the same crate).

**Comment by dtolnay** [comment]
I am interested in supporting this better. Currently my work codebase is `static_cast`ing to primitives and back, which works but can get pretty heavy on syntax.

#\[non_exhaustive\] isn't a direct solution even across crates because it's still UB for the enum to have any value other than the defined ones at runtime. For example the size of `#[non_exhaustive] enum E { Only }` is 0 bytes because there is only one possible runtime value.

**Comment by jgalenson** [comment]
In that case, should we go with a conversion function (at least for now)?

**Comment by dtolnay** [comment]
I think the most appropriate correspondence would be like this:

```cpp
enum class SmallPrime: uint16_t {
  Two = 2,
  Three = 3,
  Five = 5,
};
```

```rust
#[derive(Copy, Clone, PartialEq, Eq)]
#[repr(transparent)]
pub struct SmallPrime(u16);

#[allow(non_upper_case_globals)]
impl SmallPrime {
    pub const Two: Self = SmallPrime(2);
    pub const Three: Self = SmallPrime(3);
    pub const Five: Self = SmallPrime(5);
}
```

This works exactly like a nonexhaustive enum but matches the foreign representation.

```rust
fn demo(p: SmallPrime) {
    if p == SmallPrime::Five {
        /* ... */
    }

    match p {
        SmallPrime::Two => {}
        SmallPrime::Three => {}
        SmallPrime::Five => {}
        _other => {}
    }
}
```

**Comment by dtolnay** [comment]
Implemented in #170.
